spv/sync.go

```diff
@@ -0,0 +1,1416 @@
+			if haveBlock {
+				continue
+			}
+			log.Infof("[%d] Received sidechain or orphan block %v, height %v", key, n.Hash, n.Header.Height)
+		}
+	}
+
+	return nil
+}
+
+// hashStop is a zero value stop hash for fetching all possible data using
+// locators.
+var hashStop chainhash.Hash
+
+// getHeaders iteratively fetches headers from rp using the latest locators.
+// Returns when no more headers are available.  A sendheaders message is pushed
+// to the peer when there are no more headers to fetch.
+func (s *Syncer) getHeaders(ctx context.Context, rp *p2p.RemotePeer) error {
+
+	_, _, lowestChainWallet := s.lowestChainTip(ctx)
+
+	var locators []*chainhash.Hash
+	var generation uint
+	var err error
+	s.locatorMu.Lock()
+	locators = s.currentLocators
+	generation = s.locatorGeneration
+	if len(locators) == 0 {
+		locators, err = lowestChainWallet.BlockLocators(ctx, nil)
+		if err != nil {
+			s.locatorMu.Unlock()
+			return err
+		}
+		s.currentLocators = locators
+		s.locatorGeneration++
+	}
+	s.locatorMu.Unlock()
+
+	var lastHeight int32
+
+	for {
+		headers, err := rp.Headers(ctx, locators, &hashStop)
+		if err != nil {
+			return err
+		}
+
+		if len(headers) == 0 {
+			// Ensure that the peer provided headers through the height
+			// advertised during handshake.
+			if lastHeight < rp.InitialHeight() {
+				// Peer may not have provided any headers if our own locators
+				// were up to date.  Compare the best locator hash with the
+				// advertised height.
+				h, err := lowestChainWallet.BlockHeader(ctx, locators[0])
+				if err == nil && int32(h.Height) < rp.InitialHeight() {
+					return errors.E(errors.Protocol, "peer did not provide "+
+						"headers through advertised height")
+				}
+			}
+
+			return rp.SendHeaders(ctx)
+		}
+
+		lastHeight = int32(headers[len(headers)-1].Height)
+
+		nodes := make([]*wallet.BlockNode, len(headers))
+		g, ctx := errgroup.WithContext(ctx)
+		for i := range headers {
+			i := i
+			g.Go(func() error {
+				header := headers[i]
+				hash := header.BlockHash()
+				filter, err := rp.CFilter(ctx, &hash)
+				if err != nil {
+					return err
+				}
+				nodes[i] = wallet.NewBlockNode(header, &hash, filter)
+				return nil
+			})
+		}
+		err = g.Wait()
+		if err != nil {
+			return err
+		}
+
+		for walletID, w := range s.wallets {
+			var added int
+			s.sidechainMu.Lock()
+			for _, n := range nodes {
+				haveBlock, _, _ := w.BlockInMainChain(ctx, n.Hash)
+				if haveBlock {
+					continue
+				}
+				if s.sidechains.AddBlockNode(n) {
+					added++
+				}
+			}
+			if added == 0 {
+				s.sidechainMu.Unlock()
+
+				s.locatorMu.Lock()
+				if s.locatorGeneration > generation {
+					locators = s.currentLocators
+				}
+				if len(locators) == 0 {
+					locators, err = w.BlockLocators(ctx, nil)
+					if err != nil {
+						s.locatorMu.Unlock()
+						return err
+					}
+					s.currentLocators = locators
+					s.locatorGeneration++
+					generation = s.locatorGeneration
+				}
+				s.locatorMu.Unlock()
```

Comment:
All of this is redundant because this updated locator values are overriden below, just before the end of the loop body. The `locators` are only used when the parent infinite loop repeats, so setting them here have no relevance.